### ISSUE TITLE
Merlin container timeouts

### ISSUE THREAD

**Issue by camargo** [NASA-AMMOS/aerie#1044]
### Checked for duplicates

Yes - I've already checked

### Is this a regression?

No - This is a new bug

### Version

1.6.2

### Describe the bug

Per @parkerabercrombie We are seeing timeouts from the aerie-merlin container connecting to the database.

https://jpl.slack.com/archives/C0163E42UBF/p1690397372861419

### Reproduction

We need a good repro case for this so we can write a test to catch it.

### Logs

```shell
2023-07-26T11:45:05-07:00	[qtp370475881-1967] WARN io.javalin.Javalin - Uncaught exception
2023-07-26T11:45:05-07:00	gov.nasa.jpl.aerie.merlin.server.remotes.postgres.DatabaseException: Failed to retrieve mission model with id `7`
2023-07-26T11:45:05-07:00		at gov.nasa.jpl.aerie.merlin.server.remotes.postgres.PostgresMissionModelRepository.getMissionModel(PostgresMissionModelRepository.java:54)
2023-07-26T11:45:05-07:00		at gov.nasa.jpl.aerie.merlin.server.services.LocalMissionModelService.loadMissionModelType(LocalMissionModelService.java:285)
2023-07-26T11:45:05-07:00		at gov.nasa.jpl.aerie.merlin.server.services.LocalMissionModelService.validateActivityArguments(LocalMissionModelService.java:124)
2023-07-26T11:45:05-07:00		at gov.nasa.jpl.aerie.merlin.server.http.MerlinBindings.validateActivityArguments(MerlinBindings.java:229)
2023-07-26T11:45:05-07:00		at io.javalin.routing.HandlerEntry.handle(HandlerEntry.kt:19)
2023-07-26T11:45:05-07:00		at io.javalin.http.servlet.DefaultTasks.HTTP$lambda-8$lambda-6$lambda-5(DefaultTasks.kt:38)
2023-07-26T11:45:05-07:00		at io.javalin.http.servlet.JavalinServlet.handleTask(JavalinServlet.kt:86)
2023-07-26T11:45:05-07:00		at io.javalin.http.servlet.JavalinServlet.handleSync(JavalinServlet.kt:53)
2023-07-26T11:45:05-07:00		at io.javalin.http.servlet.JavalinServlet.service(JavalinServlet.kt:41)
2023-07-26T11:45:05-07:00		at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:587)
2023-07-26T11:45:05-07:00		at io.javalin.jetty.JavalinJettyServlet.service(JavalinJettyServlet.kt:58)
2023-07-26T11:45:05-07:00		at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:587)
2023-07-26T11:45:05-07:00		at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:764)
2023-07-26T11:45:05-07:00		at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:529)
2023-07-26T11:45:05-07:00		at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:221)
2023-07-26T11:45:05-07:00		at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1571)
2023-07-26T11:45:05-07:00		at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:221)
2023-07-26T11:45:05-07:00		at io.javalin.jetty.JettyServer$start$wsAndHttpHandler$1.doHandle(JettyServer.kt:57)
2023-07-26T11:45:05-07:00		at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
2023-07-26T11:45:05-07:00		at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:484)
2023-07-26T11:45:05-07:00		at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1544)
2023-07-26T11:45:05-07:00		at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
2023-07-26T11:45:05-07:00		at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1302)
2023-07-26T11:45:05-07:00		at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
2023-07-26T11:45:05-07:00		at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:173)
2023-07-26T11:45:05-07:00		at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
2023-07-26T11:45:06-07:00		at org.eclipse.jetty.server.HttpChannel.lambda$handle$0(HttpChannel.java:505)
2023-07-26T11:45:06-07:00		at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:762)
2023-07-26T11:45:06-07:00		at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:497)
2023-07-26T11:45:06-07:00		at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
2023-07-26T11:45:06-07:00		at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
2023-07-26T11:45:06-07:00		at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
2023-07-26T11:45:06-07:00		at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
2023-07-26T11:45:06-07:00		at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:421)
2023-07-26T11:45:06-07:00		at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:390)
2023-07-26T11:45:06-07:00		at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:277)
2023-07-26T11:45:06-07:00		at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.lambda$new$0(AdaptiveExecutionStrategy.java:139)
2023-07-26T11:45:06-07:00		at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:411)
2023-07-26T11:45:06-07:00		at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:933)
2023-07-26T11:45:06-07:00		at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1077)
2023-07-26T11:45:06-07:00		at java.base/java.lang.Thread.run(Unknown Source)
2023-07-26T11:45:06-07:00	Caused by: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30000ms.
2023-07-26T11:45:06-07:00		at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:696)
2023-07-26T11:45:06-07:00		at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:181)
2023-07-26T11:45:06-07:00		at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:146)
2023-07-26T11:45:06-07:00		at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:100)
2023-07-26T11:45:06-07:00		at gov.nasa.jpl.aerie.merlin.server.remotes.postgres.PostgresPlanRepository.getPlanRevisionData(PostgresPlanRepository.java:169)
2023-07-26T11:45:06-07:00		... 42 more
```

### System Info

```shell
Unknown
```

### Severity

Major

**Comment by parkerabercrombie** [comment]
This occurred twice during our test yesterday. We saw this in the log:
```
Caused by: com.impossibl.postgres.jdbc.PGSQLSimpleException: deadlock detected
```
In one case we had to restart the database and merlin worker containers, which resolved the situation. When this occurred there were 80 database connections active (though not 80 users, so I assume there must be more than one connection per user). It reoccurred a couple hours later, but seemed to fix itself (after a few minutes where database connectivity was severely degraded).

In both cases several users were attempting to run simulations. All the plans were branches of a single parent. Users may also have been checking constraints.

**Comment by parkerabercrombie** [comment]
Log showing deadlock
[Aerie Merlin Logs-logs-2023-07-27 08_14_51.txt](https://github.com/NASA-AMMOS/aerie/files/12185048/Aerie.Merlin.Logs-logs-2023-07-27.08_14_51.txt)

**Comment by camargo** [comment]
Potentially related to https://github.com/NASA-AMMOS/aerie/pull/661, but needs further investigation

**Comment by Mythicaeda** [comment]
This issue may have been resolved. 

The following load test suite was able to lock up the DB in the "transaction in progress" error message on 1.6.2 while using the Clipper Model (note that `get_effective_model_args` has so many users in order to have the same load on the Merlin server as a 50-50 split between it and `get_effective_activity_arguments_bulk`, which did not exist in 1.6.2):

<details closed><summary>1.6.2 Load Test Suite</summary>

```ts
scenarios: {
    insert_plans: {
      // This defines the function that this scenario will run
      exec: 'insert_plans',

      // "VUs" (virtual users) are threads that run the function we referenced above in a loop, simulating real users.
      // Pre-allocating VUs simply saves on run-time resources
      preAllocatedVUs: 50,

      // When the executor is `ramping-arrival-rate`, the stages define the number of requests
      // per second we want to target. The k6 scheduler will then automatically scale the number of VUs
      // in order to hit that desired rate

      executor: 'ramping-arrival-rate',
      stages: [
        // Start 10 iterations per second for the first 30 seconds,
        { target: 10, duration: '30s' },
        // Then, linearly scale up to 50 iterations/second for a minute,
        { target: 50, duration: '1m' },
        // Then scale down again over 30 seconds.
        { target: 10, duration: '30s' },
      ],
    },

    run_targeted_simulations: {
      exec: 'run_targeted_simulations',
      startTime: '1m',
      executor: 'constant-vus',
      vus: 2,
      duration: "30s"
    },

    run_random_simulations: {
      exec: 'run_random_simulations',
      startTime: '10s',
      executor: 'ramping-arrival-rate',
      preAllocatedVUs: 10,
      stages: [
        { target: 5, duration: '30s' },
        { target: 10, duration: '1m' },
        { target: 5, duration: '30s' },
      ],
    },

    get_effective_model_args: {
      exec: 'get_effective_model_args',
      executor: 'ramping-arrival-rate',
      preAllocatedVUs: 200,
      stages: [
        { target: 100, duration: '30s' },
        { target: 200, duration: '1m' },
        { target: 100, duration: '30s' },
      ],
    },
  },
```
</details>

This more extreme load test suite, however, had no errors when running against v1.14.0. It differs from the last test suite by including `get_effective_args` and having it and `get_effective_model_args` be using a total of 1000 Virtual Users during the peak. Since both of these endpoints run solely in the Merlin Server and are very frequently hit by the UI, the thought is that they would be the most likely endpoints to trigger the timeouts.

<details closed><summary>1.14.0 Load Test Suite</summary>

```ts
scenarios: {
    insert_plans: {
      // This defines the function that this scenario will run
      exec: 'insert_plans',

      // "VUs" (virtual users) are threads that run the function we referenced above in a loop, simulating real users.
      // Pre-allocating VUs simply saves on run-time resources
      preAllocatedVUs: 50,

      // When the executor is `ramping-arrival-rate`, the stages define the number of requests
      // per second we want to target. The k6 scheduler will then automatically scale the number of VUs
      // in order to hit that desired rate

      // https://k6.io/docs/using-k6/scenarios/executors/
      executor: 'ramping-arrival-rate',
      stages: [
        // Start 10 iterations per second for the first 30 seconds,
        { target: 10, duration: '30s' },
        // Then, linearly scale up to 50 iterations/second for a minute,
        { target: 50, duration: '1m' },
        // Then scale down again over 30 seconds.
        { target: 10, duration: '30s' },
      ],
    },

    insert_random_activities: {
      exec: 'insert_random_activities',
      executor: 'ramping-arrival-rate',
      // This scenario will only start after 5 seconds, allowing plans to populate first
      startTime: "5s",
      preAllocatedVUs: 50,
      stages: [
        { target: 10, duration: '30s' },
        { target: 50, duration: '1m' },
        { target: 10, duration: '30s' },
      ],
    },

    insert_targeted_activities: {
      exec: 'insert_targeted_activities',
      // Here we use a different executor, which keeps VUs constant, running as many iterations as
      // possible for a given duration
      executor: 'constant-vus',
      startTime: '5s',
      vus: 50,
      duration: '30s'
    },

    run_targeted_simulations: {
      exec: 'run_targeted_simulations',
      startTime: '1m',
      executor: 'constant-vus',
      vus: 2,
      duration: "30s"
    },

    run_random_simulations: {
      exec: 'run_random_simulations',
      startTime: '10s',
      executor: 'ramping-arrival-rate',
      preAllocatedVUs: 10,
      stages: [
        { target: 5, duration: '30s' },
        { target: 10, duration: '1m' },
        { target: 5, duration: '30s' },
      ],
    },

    get_effective_args: {
      exec: 'get_effective_args',
      executor: 'ramping-arrival-rate',
      preAllocatedVUs: 500,
      stages: [
        { target: 250, duration: '30s' },
        { target: 500, duration: '1m' },
        { target: 250, duration: '30s' },
      ],
    },

    get_effective_model_args: {
      exec: 'get_effective_model_args',
      executor: 'ramping-arrival-rate',
      preAllocatedVUs: 500,
      stages: [
        { target: 250, duration: '30s' },
        { target: 500, duration: '1m' },
        { target: 250, duration: '30s' },
      ],
    },
  },
```
</details>

However, because the error I got in v1.6.2 was the DB lockup and not specifically that the Merlin Server lost connection to the DB while the DB was still up, this may only be confirming that #614 has definitely been resolved.

As a next step, I will be looking into a test suite that targets the `refreshMissionModel` events and Check Constraints endpoints, as these also take place in the Merlin Server, but interact with the DB more.

**Comment by dandelany** [comment]
We believe this has been resolved in 1.14 based on @Mythicaeda 's testing - sending to @skovati to confirm & try to reproduce on his machine to double-check.

**Comment by skovati** [comment]
Using @Mythicaeda's prescribed scenarios above, I was able to reproduce on `v1.6.2` and was unable to reproduce on `v1.14.0`! I think we can close this for now, reopening if the timeout reappears 👍
